Cubical/Experiments/IntegerMatrix.agda

```diff
@@ -0,0 +1,149 @@
+{-
+
+Normalize Integer Matrices
+
+-}
+{-# OPTIONS --safe #-}
+
+module Cubical.Experiments.IntegerMatrix where
+
+open import Cubical.Foundations.Prelude
+
+open import Cubical.Data.Nat
+open import Cubical.Data.Int
+open import Cubical.Data.FinData
+open import Cubical.Data.List
+
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.CommRing.Instances.Int
+  renaming (ℤ to ℤRing)
+
+open import Cubical.Algebra.Matrix
+open import Cubical.Algebra.Matrix.CommRingCoefficient
+open import Cubical.Algebra.IntegerMatrix.Smith
+open import Cubical.Algebra.IntegerMatrix.Diagonalization
+
+private
+  variable
+    m n : ℕ
+
+open Coefficient ℤRing
+
+-- Get divisors directly
+
+open isSmithNormal
+open Smith
+open isDiagonal
+open Diag
+
+getElemDiv : Mat m n → List ℤ
+getElemDiv M = smith M .isnormal .divs .fst
+
+getDiagDiv : Mat m n → List ℤ
+getDiagDiv M = diagonalize M .isdiag .divs .fst
+
+-- Constructing matrices
+
+makeMat2×2 : ℤ → ℤ → ℤ → ℤ → Mat 2 2
+makeMat2×2 a00 _ _ _ zero zero = a00
+makeMat2×2 _ a01 _ _ zero one  = a01
+makeMat2×2 _ _ a10 _ one  zero = a10
+makeMat2×2 _ _ _ a11 one  one  = a11
+
+makeMat3×3 : ℤ → ℤ → ℤ → ℤ → ℤ → ℤ → ℤ → ℤ → ℤ → Mat 3 3
+makeMat3×3 a00 _ _ _ _ _ _ _ _ zero zero = a00
+makeMat3×3 _ a01 _ _ _ _ _ _ _ zero one  = a01
+makeMat3×3 _ _ a02 _ _ _ _ _ _ zero two  = a02
+makeMat3×3 _ _ _ a10 _ _ _ _ _ one  zero = a10
+makeMat3×3 _ _ _ _ a11 _ _ _ _ one  one  = a11
+makeMat3×3 _ _ _ _ _ a12 _ _ _ one  two  = a12
+makeMat3×3 _ _ _ _ _ _ a20 _ _ two  zero = a20
+makeMat3×3 _ _ _ _ _ _ _ a21 _ two  one  = a21
+makeMat3×3 _ _ _ _ _ _ _ _ a22 two  two  = a22
```

Comment:
Why not use https://github.com/agda/cubical/blob/master/Cubical/Algebra/Matrix.agda#L50 ?

Reply:
I tried before. The result performs a bit slower than this direct definition, so I decided to use this one.

Reply:
Ah ok, fair enough!

Reply:
Maybe should be commented :/

Reply:
@ice1000 I'll comment it in future PR :P